src/utils.c

```diff
@@ -743,3 +743,169 @@ gboolean r_fakeroot_cleanup(const gchar *envpath, GError **error)
 
 	return rm_tree(tmpdir, error);
 }
+
+gboolean semver_parse(const gchar *version_string, guint version_core[3], gchar **pre_release, gchar **build)
+{
+	g_autofree gchar *version_copy = g_strdup(version_string);
+	gchar *build_pos = NULL;
+	gchar *pre_release_pos = NULL;
+	gchar **version_parts = NULL;
+
+	if (pre_release != NULL)
+		*pre_release = NULL;
+
+	if (build != NULL)
+		*build = NULL;
+
+	/* semantic versions BNF:
+	 * <valid semver> ::= <version core>
+	 * | <version core> "-" <pre-release>
+	 * | <version core> "+" <build>
+	 * | <version core> "-" <pre-release> "+" <build>
+	 *
+	 * detect first the '+build', separate it out by null terminating the
+	 * version_copy at it's start position; repeat for the '-pre-release'; and
+	 * then parse the remaining 'version core'
+	 */
+
+	build_pos = g_strrstr(version_copy, "+");
+	if (build_pos != NULL) {
+		if (build != NULL) {
+			*build = g_strdup(build_pos + 1);
+			if (*build == NULL)
+				goto out;
+		}
+		*build_pos = '\0';
+	}
+
+	pre_release_pos = g_strrstr(version_copy, "-");
+	if (pre_release_pos != NULL) {
+		if (pre_release != NULL) {
+			*pre_release = g_strdup(pre_release_pos + 1);
+			if (*pre_release == NULL)
+				goto out;
+		}
+		*pre_release_pos = '\0';
+	}
+
+	version_parts = g_strsplit(version_copy, ".", 3);
+	// Note that g_strsplit returns an empty array when splitting: ""
+	if (version_parts == NULL)
+		goto out;
+	// and NULL terminates arrays it creates
+	if (version_parts[0] == NULL)
+		goto out;
+	// only on component -> shift right
+	if (version_parts[1] == NULL) {
+		version_parts[2] = version_parts[0];
+		version_parts[0] = NULL;
+	}
+	// only two components -> shift right
+	if (version_parts[1] != NULL && version_parts[2] == NULL) {
+		version_parts[2] = version_parts[1];
+		version_parts[1] = version_parts[0];
+		version_parts[0] = NULL;
+	}
+
+	for (int i = 0; i < 3; i++) {
+		if (version_parts[i] != NULL) {
+			// Note that negative numbers are already implicitly excluded due to the search for the '-pre_release'
+			version_core[i] = g_ascii_strtoll(version_parts[i], NULL, 10);
+			// handle g_ascii_strtoll returning zero on a conversion error
+			if ((version_core[i] == 0) && (g_strcmp0(version_parts[i], "0") != 0)) {
+				g_strfreev(version_parts);
+				goto out;
+			}
+		} else
+			version_core[i] = 0;
+	}
+
+	g_strfreev(version_parts);
+	return TRUE;
+
+out:
+	if (pre_release && *pre_release)
+		g_free(*pre_release);
+	if (build && *build)
+		g_free(*build);
+	return FALSE;
+}
+
+gboolean semver_less_equal(const guint version_a[3], gchar *pre_release_a, const guint version_b[3], gchar *pre_release_b)
+{
+	gchar **pre_fields_a = NULL;
+	gchar **pre_fields_b = NULL;
+	gboolean is_num_a = FALSE;
+	gboolean is_num_b = FALSE;
+	guint num_a = 0;
+	guint num_b = 0;
+	gint cmp = 0;
```

Comment:
Move these declarations into the loop scopes.

Reply:
ACK, done with v3